kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -311,29 +478,79 @@ public void initiateConnect(HostPort remote, List<FilterAndInvoker> filters) {
+        final Future<Channel> handshakeFuture = sslHandshakeFuture.get();
+        Future<?> kafkaAvailableFuture = handshakeFuture != null ? handshakeFuture : tcpConnectFuture;
+        kafkaAvailableFuture.addListener(future -> {
             if (future.isSuccess()) {
-                state = State.CONNECTED;
-                LOGGER.trace("{}: Outbound connected", inboundCtx.channel().id());
-                // Now we know which filters are to be used we need to update the DecodePredicate
-                // so that the decoder starts decoding the messages that the filters want to intercept
-                dp.setDelegate(DecodePredicate.forFilters(filters));
+                onConnection(filters);
             }
             else {
-                state = State.FAILED;
-                // Close the connection if the connection attempt has failed.
-                Throwable failureCause = future.cause();
-                LOGGER.atWarn()
-                        .setCause(LOGGER.isDebugEnabled() ? failureCause : null)
-                        .log("Connection to target cluster on {} failed with: {}, closing inbound channel. Increase log level to DEBUG for stacktrace", remote,
-                                failureCause.getMessage());
-                inboundChannel.close();
+                onConnectionFailed(remote, future, inboundChannel);
             }
         });
     }
 
+    private @NonNull Object buildErrorResponseFrame(DecodedRequestFrame<?> triggerFrame, Throwable error) {
+        var responseData = errorResponse(triggerFrame, error);
+        final ResponseHeaderData responseHeaderData = new ResponseHeaderData();
+        responseHeaderData.setCorrelationId(triggerFrame.correlationId());
+        return new DecodedResponseFrame<>(triggerFrame.apiVersion(), triggerFrame.correlationId(), responseHeaderData, responseData);
+    }
+
+    @VisibleForTesting
+    void onConnectionFailed(HostPort remote, Future<?> future, Channel inboundChannel) {
+        state = State.FAILED;
+        // Close the connection if the connection attempt has failed.
+        Throwable failureCause = future.cause();
+        handleException(failureCause).ifPresentOrElse(
+                result -> closeWith(inboundChannel, result),
+                () -> {
+                    LOGGER.atWarn()
+                            .setCause(LOGGER.isDebugEnabled() ? failureCause : null)
+                            .log("Connection to target cluster on {} failed with: {}, closing inbound channel. Increase log level to DEBUG for stacktrace",
+                                    remote, failureCause.getMessage());
+                    closeOnFlush(inboundChannel);
+                });
+    }
+
+    @SuppressWarnings("java:S1452")
+    Optional<?> handleException(Throwable throwable) {
+        var localCause = throwable;
+        while (localCause != null) {
```

Comment:
Seeing this code brought back memories of LOGBACK-1027.   That was talking about suppressed exceptions (which I don't think we need to worry about).

But it does make me nervous about cause loops leading to an infinite loop.

`#printStackTrace` uses code like this to keep track of the exceptions it has seen.
```
        Set<Throwable> dejaVu = Collections.newSetFromMap(new IdentityHashMap<>());
        dejaVu.add(this);
```